xarray_beam/_src/pangeo.py

```diff
@@ -0,0 +1,150 @@
+
+  return dataset
+
+
+class FilePatternToChunks(beam.PTransform):
+  """Open data described by a Pangeo-Forge `FilePattern` into keyed chunks."""
+
+  from pangeo_forge_recipes.patterns import FilePattern
+
+  def __init__(
+      self,
+      pattern: 'FilePattern',
+      sub_chunks: Optional[Mapping[str, int]] = None,
+      xarray_open_kwargs: Optional[Dict] = None
+  ):
+    """Initialize FilePatternToChunks.
+
+    TODO(#29): Currently, `MergeDim`s are not supported.
+
+    Args:
+      pattern: a `FilePattern` describing a dataset.
+      sub_chunks: split each open dataset into smaller chunks. If not set, each
+        chunk will open the full dataset.
+      xarray_open_kwargs: keyword arguments to pass to `xarray.open_dataset()`.
+    """
+    self.pattern = pattern
+    self.sub_chunks = sub_chunks or -1
+    self.xarray_open_kwargs = xarray_open_kwargs or {}
+
+  def _prechunk(self) -> Iterator[Tuple[core.Key, Tuple[int, ...]]]:
+    """Converts `FilePattern` items into keyed indexes."""
+    input_chunks = {k: v or 1 for k, v in self.pattern.nitems_per_input.items()}
+    dim_sizes = {
+      k: v or self.pattern.dims[k]
+      for k, v, in self.pattern.concat_sequence_lens.items()
+    }
+    chunks = core.normalize_expanded_chunks(input_chunks, dim_sizes)
+    for key, (index, _) in zip(core.iter_chunk_keys(chunks),
+                               self.pattern.items()):
+      yield key, index
+
+  def _open_chunks(
+      self,
+      key: core.Key,
+      index: Tuple[int, ...]
+  ) -> Iterator[Tuple[core.Key, xarray.Dataset]]:
+    """Open datasets into chunks with XArray."""
+    path = self.pattern[index]
+    with FileSystems().open(path) as file:
+      dataset = xarray.open_dataset(
+        file, chunks=self.sub_chunks, **self.xarray_open_kwargs
+      )
+      dataset = _expand_dimensions_by_key(dataset, key, index, self.pattern)
+
+      base_key = core.Key(_zero_dimensions(dataset)).with_offsets(**key.offsets)
+
+      num_threads = len(dataset.data_vars)
+
+      if self.sub_chunks == -1:
+        yield base_key, dataset.compute(num_workers=num_threads)
+        return
+
+      dim_sizes = {dim: dataset.dims[dim] for dim in self.sub_chunks.keys()}
+      norm_sub_chunks = core.normalize_expanded_chunks(
+        self.sub_chunks, cast(Mapping[str, int], dim_sizes)
+      )
+      offset_index = core.compute_offset_index(
+        core._chunks_to_offsets(norm_sub_chunks)
+      )
+      for sub_key in core.iter_chunk_keys(norm_sub_chunks):
+        sizes = {
+          dim: norm_sub_chunks[dim][offset_index[dim][offset]]
+          for dim, offset in sub_key.offsets.items()
+        }
+        slices = core.offsets_to_slices(
+          sub_key.offsets, sizes=sizes, base=base_key.offsets
+        )
+        chunk = dataset.isel(slices)
+
+        new_key = base_key.with_offsets(**sub_key.offsets)
+        yield new_key, chunk.chunk().compute(num_workers=num_threads)
```

Comment:
Could this be rewritten in terms of `rechunk.split_chunks` instead? My main interest is to try to reduce duplicate indexing/rechunking logic.

Something like:
```python
for new_key, chunk in rechunk.split_chunks(
     key, dataset, self.sub_chunks,
):
  yield new_key, chunk.compute(...)
```

The only real difference here is that what you have here (like the existing `DatasetToChunks`) handles expanded _tuples_ of chunks, e.g., `{'x': (3, 3, 3, 1)}` vs just integer values, e.g., `{'x': 3}`.

The main reason why I supported expanded tuples is because it's what `xarray.Dataset.chunks` provides. It's theoretically more flexible than a single chunk size, but I doubt users will ever want to supply that explicitly themselves.

Reply:
I'd be much happier with simpler code here; when I write this, I don't think I was familiar with `rechunk.split_chunks`.

Reply:
Your example works right out of the box! It feels good to delete so much code :)